src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -137,94 +144,177 @@ boolean removeUnreachableBlocks ()
-	 * subset, which consists of those which have side-effect or produce a value
-	 * consumed by other needed instructions.  Don't assume SSA.
+	 * Remove any unnecessary instructions.  Answer true if any were removed.
 	 *
-	 * @return The instructions that are needed and should be kept.
+	 * @param dataCouplingMode
+	 *        The {@link DataCouplingMode} that chooses how to trace liveness.
+	 * @return Whether any dead instructions were removed or changed.
 	 */
-	private Set<L2Instruction> findNeededInstructions ()
+	private boolean removeDeadInstructions (
+		final DataCouplingMode dataCouplingMode)
 	{
-		final Deque<L2Instruction> instructionsToVisit = new ArrayDeque<>();
+		final DeadCodeAnalyzer analyzer =
+			new DeadCodeAnalyzer(dataCouplingMode, controlFlowGraph);
+		analyzer.analyzeReads();
+		final Set<L2Instruction> liveInstructions = analyzer.liveInstructions();
+		boolean anyRemoved = false;
 		for (final L2BasicBlock block : blocks)
 		{
-			for (final L2Instruction instruction : block.instructions())
+			final ListIterator<L2Instruction> iterator =
+				block.instructions().listIterator();
+			while (iterator.hasNext())
 			{
-				if (instruction.hasSideEffect())
+				final L2Instruction instruction = iterator.next();
+				if (!liveInstructions.contains(instruction))
 				{
-					instructionsToVisit.add(instruction);
+					anyRemoved = true;
+					final @Nullable L2Instruction replacement =
+						instruction.optionalReplacementForDeadInstruction();
+
+					if (replacement == null)
+					{
+						iterator.remove();
+						instruction.justRemoved();
+					}
+					else
+					{
+						iterator.set(replacement);
+						instruction.justRemoved();
+						if (replacement.operation() == L2_JUMP.instance)
+						{
+							replacement.justInserted();
+						}
+					}
 				}
 			}
 		}
-		// Recursively mark as needed all instructions that produce values
-		// consumed by another needed instruction.
-		final Set<L2Instruction> neededInstructions = new HashSet<>();
-		while (!instructionsToVisit.isEmpty())
+		if (anyRemoved)
 		{
-			final L2Instruction instruction = instructionsToVisit.removeLast();
-			if (!neededInstructions.contains(instruction))
-			{
-				neededInstructions.add(instruction);
-				for (final L2Register sourceRegister
-					: instruction.sourceRegisters())
-				{
-					// Assume all definitions are needed, regardless of control
-					// flow.
-					instructionsToVisit.addAll(sourceRegister.definitions());
-				}
-			}
+			updateAllSemanticValuesOfReads();
+			updateAllManifests();
 		}
-		return neededInstructions;
+		return anyRemoved;
 	}
 
 	/**
-	 * Remove any unnecessary instructions.  Answer true if any were removed.
+	 * Visit all manifests, retaining only the {@link L2Register}s that are
+	 * still written to by live code.
+	 */
+	private void updateAllManifests ()
+	{
+		// Visit each edge's manifest, removing from its definitions lists all
+		// writes from removed instructions.
```

Comment:
This comment is incoherent. Maybe "definition" should be singular or "lists" should be omitted completely. I can't quite tell what's happening here. Based on the code itself, it's probably cleanest to drop "lists".

Reply:
It's a bit tricky, so I separately commented the two passes.